### ISSUE TITLE
There is a small error in the gradient backword

### ISSUE THREAD

**Issue by rimu123** [generalized-iou/g-darknet#20]
Hi,@nathantsoi
Thank you for the work done. 
I found a small bug in your code. Maybe that is why the training converges slowly.Amend as below:
`delta[index + 0*stride] = (all_ious.dx_iou.dl + all_ious.dx_iou.dr)/lw; `
`delta[index + 1*stride] = (all_ious.dx_iou.dt + all_ious.dx_iou.db)/lh; `
`delta[index + 2*stride] = ((-0.5 * all_ious.dx_iou.dl) + (0.5 * all_ious.dx_iou.dr))*pred.w; `
`delta[index + 3*stride] = ((-0.5 * all_ious.dx_iou.dt) + (0.5 * all_ious.dx_iou.db))*pred.h;`
Compared with:
`delta[index + 0*stride] = (all_ious.dx_iou.dl + all_ious.dx_iou.dr);`
` delta[index + 1*stride] = (all_ious.dx_iou.dt + all_ious.dx_iou.db); `
`delta[index + 2*stride] = ((-0.5 * all_ious.dx_iou.dl) + (0.5 * all_ious.dx_iou.dr)); `
`delta[index + 3*stride] = ((-0.5 * all_ious.dx_iou.dt) + (0.5 * all_ious.dx_iou.db));`
 because of:
`b.x = (i + x[index + 0stride]) / lw;`
`b.y = (j + x[index + 1stride]) / lh;`
`b.w = exp(x[index + 2stride]) * biases[2n] / w;`
`b.h = exp(x[index + 3stride]) * biases[2n+1] / h;`
According to the chain rulem, the gradient needs to be multiplied by the expression of x,y,w,h. It works well on my training set，rised by 3 points. By a way, due to the countdown, the gradient of gIOU is not perfect. I hope that can give you some help.

**Comment by daixiaoxi** [comment]
> Hi,@nathantsoi
> Thank you for the work done.
> I found a small bug in your code. Maybe that is why the training converges slowly.Amend as below:
> delta[index + 0*stride] = (all_ious.dx_iou.dl + all_ious.dx_iou.dr)/lw; 
> delta[index + 1*stride] = (all_ious.dx_iou.dt + all_ious.dx_iou.db)/lh; 
> delta[index + 2*stride] = ((-0.5 * all_ious.dx_iou.dl) + (0.5 * all_ious.dx_iou.dr))*pred.w; 
> delta[index + 3*stride] = ((-0.5 * all_ious.dx_iou.dt) + (0.5 * all_ious.dx_iou.db))*pred.h;
> Compared with:
> delta[index + 0*stride] = (all_ious.dx_iou.dl + all_ious.dx_iou.dr);
> delta[index + 1*stride] = (all_ious.dx_iou.dt + all_ious.dx_iou.db);
> delta[index + 2*stride] = ((-0.5 * all_ious.dx_iou.dl) + (0.5 * all_ious.dx_iou.dr)); 
> delta[index + 3*stride] = ((-0.5 * all_ious.dx_iou.dt) + (0.5 * all_ious.dx_iou.db));
> because of:
> b.x = (i + x[index + 0stride]) / lw;
> b.y = (j + x[index + 1stride]) / lh;
> b.w = exp(x[index + 2stride]) * biases[2n] / w;
> b.h = exp(x[index + 3stride]) * biases[2n+1] / h;
> According to the chain rulem, the gradient needs to be multiplied by the expression of x,y,w,h. It works well on my training set，rised by 3 points. By a way, due to the countdown, the gradient of gIOU is not perfect. I hope that can give you some help.



> Hi,@nathantsoi
> Thank you for the work done.
> I found a small bug in your code. Maybe that is why the training converges slowly.Amend as below:
> delta[index + 0*stride] = (all_ious.dx_iou.dl + all_ious.dx_iou.dr)/lw; 
> delta[index + 1*stride] = (all_ious.dx_iou.dt + all_ious.dx_iou.db)/lh; 
> delta[index + 2*stride] = ((-0.5 * all_ious.dx_iou.dl) + (0.5 * all_ious.dx_iou.dr))*pred.w; 
> delta[index + 3*stride] = ((-0.5 * all_ious.dx_iou.dt) + (0.5 * all_ious.dx_iou.db))*pred.h;
> Compared with:
> delta[index + 0*stride] = (all_ious.dx_iou.dl + all_ious.dx_iou.dr);
> delta[index + 1*stride] = (all_ious.dx_iou.dt + all_ious.dx_iou.db);
> delta[index + 2*stride] = ((-0.5 * all_ious.dx_iou.dl) + (0.5 * all_ious.dx_iou.dr)); 
> delta[index + 3*stride] = ((-0.5 * all_ious.dx_iou.dt) + (0.5 * all_ious.dx_iou.db));
> because of:
> b.x = (i + x[index + 0stride]) / lw;
> b.y = (j + x[index + 1stride]) / lh;
> b.w = exp(x[index + 2stride]) * biases[2n] / w;
> b.h = exp(x[index + 3stride]) * biases[2n+1] / h;
> According to the chain rulem, the gradient needs to be multiplied by the expression of x,y,w,h. It works well on my training set，rised by 3 points. By a way, due to the countdown, the gradient of gIOU is not perfect. I hope that can give you some help.

Hello, I use the backword. When I debug the code ,I found the delta[index + 0*stride] and the delta[index + 1*stride] always to be zero. It means that all_ious.dx_iou.dl and all_ious.dx_iou.dr are opposite. Have you ever meet the problem ? Thanks a lot

**Comment by rimu123** [comment]
Training a model with the original darknet and then fine-tuning with gIOU may get a good result.
